messenger/src/pulsar_messenger.rs

```diff
@@ -0,0 +1,237 @@
+                });
+            }
+        }
+
+        // Check if topic is configured
+        let producer = if let Some(producer) = self.producers.get_mut(stream_key) {
+            producer
+        } else {
+            error!("Cannot send data for topic {stream_key}, it is not configured");
+            return Err(MessengerError::SendError {
+                msg: String::from(format!(
+                    "Cannot send data for topic {:?}, it is not configured",
+                    stream_key
+                )),
+            });
+        };
+
+        let result = producer
+            .send(bytes)
+            .await
+            .map_err(|e| MessengerError::SendError {
+                msg: String::from(format!(
+                    "Error while sending message to the Pulsar: {:?}",
+                    e
+                )),
+            })?
+            .await;
+
+        if let Err(e) = result {
+            error!("Did not get message receipt: {e}");
+            return Err(MessengerError::SendError { msg: e.to_string() });
+        } else {
+            info!("Data sent");
+        }
+
+        Ok(())
+    }
+
+    /// Receive message from the Pulsar topic
+    async fn recv(
+        &mut self,
+        stream_key: &'static str,
+    ) -> Result<Vec<(i64, &[u8])>, MessengerError> {
+        // Check if consumer is exists
+        let mut consumer = if let Some(consumer) = self.consumers.get(stream_key) {
+            consumer.lock().await
+        } else {
+            error!("Cannot get data from topic {stream_key}, consumer is not configured");
+            return Err(MessengerError::ReceiveError {
+                msg: String::from("Consumer for the requested topic wasn't created"),
+            });
+        };
+
+        let next_message = consumer
+            .try_next()
+            .await
+            .map_err(|e| MessengerError::ReceiveError { msg: e.to_string() })?;
+
+        if let Some(msg) = next_message {
+            if let Some(messages) = self.acquired_messages.get_mut(stream_key) {
+                messages.push(msg);
+            } else {
+                self.acquired_messages.insert(stream_key, vec![msg]);
+            }
+            let data = self
+                .acquired_messages
+                .get(stream_key)
+                .unwrap()
+                .last()
+                .unwrap()
+                .payload
+                .data
+                .as_ref(); // safe to use unwrap because we just pushed data to the HashMap
+
+            return Ok(vec![(0, data)]); // TODO: it is not universal data type
+        } else {
+            return Err(MessengerError::ReceiveError {
+                msg: String::from("No data in requested topic found"),
+            });
+        }
+    }
+}
+
+/// Acknowledge reading message from the Pulsar by removing it from the HashMap and calling Pulsar.ack()
+/// Should be called after `recv()` method
+pub async fn ack_pulsar_message(
```

Comment:
I know we already approved this PR but something I thought of we should discuss for this or can be a follow-on PR.   I think `ack_recv()` should be a method added to the `Messenger` trait, even though the `RedisMessenger` would probably not do anything with it.

I think this would help with the goal of the `Messenger` trait, that applications can mostly use the generic `Messenger` methods, and then changing out the specific `Messenger` should require minimal or no changes to application code.  So then we would add the `Messenger` `recv_ack`s to the right spots in `nft_ingestor` as well.